src/verilator/verilator_object.cc

```diff
@@ -0,0 +1,92 @@
+#include "base/logging.hh"
+#include "sim/sim_exit.hh"
+#include "verilator_object.hh"
+
+VerilatorObject::VerilatorObject(VerilatorObjectParams *params) :
+    SimObject(params),
+    event([this]{processEvent();}, params->name),
+    maxCycles(params->cycles),
+    latency(params->latency),
+   designStages(params->stages),
+   start(params->startTime),
+    loadMem(params->memData.c_str()),
+   objName(params->name)
+{
+    //send program or other data to fesvr to communicate
+    //with design
+    toDtm.push_back(loadMem);
+    if (!toDtm.size()){
+        //DPRINT
+        fatal("No binary specified for emulator\n");
+    }
+
+    inform("Instantiated DTM.\n");
+    //give fesvr list of things to load/parse
+    dtm = new dtm_t(toDtm);
+    if (!dtm){
+        panic("Could not allocate dtm\n");
+    }
+}
+
+VerilatorObject::~VerilatorObject()
+{
+    delete dtm;
+}
+
+VerilatorObject*
+VerilatorObjectParams::create()
+{
+    VerilatorObject * vo = new VerilatorObject(this);
+    return vo;
+}
+
+void
+VerilatorObject::processEvent()
+{
+    //has verilator finished running?
+    if (dtm->done() && dut.io_success && Verilated::gotFinish()){
```

Comment:
`dut.io_success` is specific to the sodor CPU. I think this file shouldn't have any model-specific stuff in it. Verilated::gotFinish() is good here since it's available for all verilated files.